Prevent set_pipeline runtime error

mdb opened the pull request:
## What does this PR accomplish?

This PR seeks to address issue #6111 by preventing a `runtime error: invalid memory address or nil pointer dereference` when `set_pipeline` is run with a `file:` whose YML contains unexpected content.

## Changes proposed by this PR:

This PR attempts to prevent the runtime error in the most minimal way possible by _only_ attempting to invoke `pipeline.SetParentIDs(step.metadata.JobID, step.metadata.BuildID)` _if_ a pipeline is found by `team.Pipeline(pipelineRef)`, in effect avoiding a scenario where `pipeline` doesn't exist and where `pipeline.SetParentIDs()` causes a runtime error.

One could imagine other solutions, such as perhaps improved [configvalidate#Validate](https://github.com/concourse/concourse/blob/master/atc/configvalidate/validate.go#L25) functionality to protect against this. However, the solution implemented in this PR aspires to be a simple first pass at fixing the problem with a minimal code change.

## Notes to reviewer:

This fix can be verified by running `docker-compose up --build` against this code and `fly set-pipeline`-ing the following pipeline...

```yaml
resource_types:

- name: git-ubuntu
  type: registry-image
  source:
    repository: concourse/git-resource
    tag: ubuntu

resources:

- name: gist
  type: git-ubuntu
  source:
    uri: https://gist.github.com/ac6ae0980b268ae5ff2449fef3ab87c5.git

jobs:

- name: update-pipeline-with-task-yml
  plan:
  - get: gist
  - set_pipeline: bad-pipeline
    file: gist/task.yml

- name: update-pipeline-with-empty-yml
  plan:
  - get: gist
  - set_pipeline: bad-pipeline
    file: gist/empty.yml

- name: update-pipeline-with-valid-yml
  plan:
  - get: gist
  - set_pipeline: good-pipeline
    file: gist/valid-pipeline.yml
```

Running `update-pipeline-with-task-yml` and `update-pipeline-with-empty-yml` should not result in newly-set `bad-pipeline` pipelines. Running `update-pipeline-with-valid-yml` _should_ result in a newly-set `good-pipeline`.

## Release Note

`set_pipeline` of a YML pipeline configuration file with no `jobs:` or `resources:` no longer causes a `runtime error: invalid memory address or nil pointer dereference`.

## Contributor Checklist

- [X] Followed [Code of conduct], [Contributing Guide] & avoided [Anti-patterns]
- [X] [Signed] all commits
- [X] Added tests (Unit and/or Integration)
- [ ] Updated [Documentation]
- [ ] Added release note (Optional)

[Code of Conduct]: https://github.com/concourse/concourse/blob/master/CODE_OF_CONDUCT.md
[Contributing Guide]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md
[Anti-patterns]: https://github.com/concourse/concourse/wiki/Anti-Patterns
[Signed]: https://help.github.com/en/github/authenticating-to-github/signing-commits
[Documentation]: https://github.com/concourse/docs

## Reviewer Checklist
<!--
This section is intended for the reviewers only, to track review
progress.
-->
- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed
- [ ] New config flags added? Ensure that they are added to the
  [BOSH](https://github.com/concourse/concourse-bosh-release) and
  [Helm](https://github.com/concourse/helm) packaging; otherwise, ignored for
  the [integration
  tests](https://github.com/concourse/ci/tree/master/tasks/scripts/check-distribution-env)
  (for example, if they are Garden configs that are not displayed in the
  `--help` text).

taylorsilva reviewed (commented):
Thanks for the PR! Just played around with it to see what the before and after looks like. 

I was surprised to see all the jobs go green; it felt misleading. With what's on master right now the jobs actually enter a failure state, not the right one, but it at least tells the user "hey something went wrong, start digging". With the jobs going green this information is now hidden and someone wouldn't realize that the job was actually a no-op until 1) they find their pipeline wasn't created or 2) their pipeline isn't reflecting their update.

Seeing as the nil pointer doesn't bring down the ATC anymore either I'd actually prefer to keep things as-is and work towards the ideal state of erroring and printing a useful message for the user. wdyt?

mdb commented:
@taylorsilva Thanks for the feedback (and it was nice meeting you on the office hours call this evening)!

> With the jobs going green this information is now hidden and someone wouldn't realize that the job was actually a no-op until 1) they find their pipeline wasn't created or 2) their pipeline isn't reflecting their update.

You make a good point. However, it's my understanding that the _reason_ no pipeline is created is because the YML results in a [atc.Config](https://github.com/concourse/concourse/blob/7f1c1ea44b38dcc94c34dbab9e6e7d767a101c7d/atc/config.go#L19-L26) whose fields (`Groups`, `VarSources`, `Resources`, `ResourceTypes`, `Jobs`, `Display`) are all empty.

So, with this in mind, I would expect `set_pipeline`-ing such a config would behave akin to its `fly set-pipeline` analog. However, `fly set-pipeline` of the same YML exits zero (success!) and reports "no changes to apply" (Is the inconsistent "no diff found" (`set_pipeline`) VS "no changes to apply" (`fly set-pipeline`) messaging perhaps another [issue](https://github.com/concourse/concourse/issues) worth documenting? Or is that intentional?):

```
./fly -t local set-pipeline --pipeline empty --config empty.yml
no changes to apply

echo $?
0

./fly -t local set-pipeline --pipeline empty --config task.yml
no changes to apply

echo $?
0
```

Furthermore `fly validate-pipeline` reports that the YML "looks good:"

```
./fly validate-pipeline --config empty.yml
looks good 

./fly validate-pipeline --config task.yml
looks good 
```

So, I would expect that `set_pipeline` and `fly set-pipeline` behave similarly when fed the same YML configuration.

But, if you'd prefer `set_pipeline` error in such a scenario, is it also worth reconsidering `fly set-pipeline` and `fly validate-pipeline`'s behavior? Such deeper reconsideration is some of what I was alluding to when I mentioned "One could imagine other solutions, such as perhaps improved `configvalidate#Validate` functionality to protect against this." However, I opted to punt on those more thorny product and code changes in favor of this relatively minimal change, which at least establishes a bit more consistency for now 😄 

Does that change your view?

taylorsilva commented:
@mdb That makes sense: fly is behaving one way currently and you're using that as the baseline for how `set_pipeline` should behave. Normally I would go with that too but I still think that brings use to an undesirable UX for `set_pipeline`. I think fly's current behaviour is mostly okay because the user does see specific feedback and can act on it if it doesn't meet their expectations (e.g. they `fly set-pipline` and get `no changes to apply` when they were expecting a diff). I think fly's behaviour could be improved too though to cover the examples mentioned in your PR and tests.

Not sure if we do this in the pipeline validation logic already but I'm thinking an easy way to see if a pipeline is valid is to check if it contains at least one job. I know I'm pulling you into another section of the codebase, but I'm thinking we could have both fly and `set_pipeline` behave the same by checking if a pipeline config contains zero jobs.

We could potentially add this logic here:

https://github.com/concourse/concourse/blob/7948c873d3e4c3949201f40e99c2570654df570c/atc/configvalidate/validate.go#L53-L58

Something like:
```go
if len(c.JobConfigs) == 0 {
    warnings = append(warnings, ConfigWarning{Type: "jobs", Message: "config contains zero jobs"})
}
```

I quickly scanned the validate test suite and didn't find anything doing this check. I think this could be an equally small change that may get us to our desired state. What do you think?

taylorsilva commented:
@vito curious about your thoughts here ^
trying to figure out if there are any gotchas here or what the historical reason is for not having this check already?

Chatted with @xtreme-sameer-vohra about it and he pointed out that `jobs` is optional according to our docs: https://concourse-ci.org/pipelines.html

The question I'm trying to answer is: what is considered the most minimal valid pipeline?
I think the answer is _a single valid job_. Is there a reason that is not true?

vito commented:
@taylorsilva Yeah, I'd say checking for a job to be present is sensible. In fact if you try to add a resource without adding a job you'll get a validation error saying the resource is unused.

Just a quick note: we expressly allow unknown toplevel keys, because folks use them to tuck away YAML `&anchors` to be referenced elsewhere in the config. So that's why weird things like a task config are valid: we ignore the unknown fields, and accept an empty pipeline.

The only practical reason I can think of to accept an empty config is if we want to allow someone to comment out the entire darn YAML and configure it. But it seems like you could also just pause the pipeline.

Beyond that, it's kind of a philosophical question: given a pipeline is just a set of N jobs, is a set of 0 jobs a valid pipeline? I think it feels kind of nice to allow it, since it's sort of the "base case" given our definition of a pipeline, but pragmatically speaking it's more likely a user error and it'd be nice to report those early - so it's not a hill I'd seek to die on. 😄

So I'd be OK with adding a validation for this, but I also think the code submitted so far is worth merging. Not checking the `found` value looks like a panic waiting to happen.

taylorsilva commented on atc/exec/set_pipeline_step_test.go:
```diff
@@ -241,6 +257,54 @@ jobs:
 			})
 		})
 
+		Context("when pipeline file exists but has task syntax", func() {
```
For the tests, these won't be valid once we update the `configvalidate` package and they don't express the desired UX. The following is weird but it is essentially the scenario that your team ran into (kinda?).

There's a context `when specified pipeline not found`, and under that could be another context that says `when no diff is found`. That'll push us down the codepath to verify that we don't call `pipeline.SetParentIDs()` when no pipeline was found.
```go
Context("when no diff is found", func() {
  ...
}

mdb replied:
@taylorsilva Thanks! I _believe_ I've modified the tests as you're describing. Please let me know if I've misinterpreted.

taylorsilva reviewed (changes requested):
Need to rebase this PR now. 
Left one comment about updating the tests and then this will be good to merge!

taylorsilva commented:
> Just a quick note: we expressly allow unknown toplevel keys, because folks use them to tuck away YAML &anchors to be referenced elsewhere in the config. So that's why weird things like a task config are valid: we ignore the unknown fields, and accept an empty pipeline.

Thanks, that's exactly what I was trying to surface. Thankfully checking there's one job should not break this in any way.

mdb commented:
@vito and @taylorsilva Thanks for the patient reviews. I've rebased, made what I believe to be the suggested test changes, and added a release note. 

While it's only peripherally related to this PR, I'm also curious to hear what you all think about this question (I'd be happy to submit a PR for it as well):

> Is the inconsistent "no diff found" (`set_pipeline`) VS "no changes to apply" (`fly set-pipeline`) messaging perhaps another issue worth documenting? Or is that intentional?

Thanks!

vito commented:
Good catch - I think we should make the set_pipeline step consistent with the fly output.

taylorsilva commented on release-notes/latest.md:
```diff
@@ -33,3 +33,7 @@
 
   A new metric called `tasks_wait_duration_bucket` is also added to express as quantiles the average time spent by tasks awaiting execution. PR: #5981
   ![Example graph for the task wait time histograms.](https://user-images.githubusercontent.com/40891147/89990749-189d2600-dc83-11ea-8fde-ae579fdb0a0a.png)
+
+#### <sub><sup><a name="6111" href="#6111">:link:</a></sup></sub> bug
+
+* @mdb Running `set_pipeline` against a pipeline config YML file with no `jobs:` no longer causes runtime error. Issue: #6111 PR: #6116
```
No need to do this. When the next version of Concourse is released we read out whatever you put in the `Release Notes` header of this PR.

mdb replied:
@taylorsilva Gotcha; I've removed this commit. I was following the instructions in the `CONTRIBUTING.md` doc. Perhaps it's worth updating these instructions?

taylorsilva reviewed (changes requested):
Remove the release notes diif and then this will be good to go!
